### PR TITLE
web: Update from bootstrap 3.3.1 to bootstrap 4.0.0

### PR DESCRIPTION
**Pull request by achiuBAE**
implemented: #4754 

- Updated the html template files to be bootstrap 4.0.0 compatible, ensuring that all the functionality and ui is the same.
vendor: 
-reimplemented glyphicons from bootstrap 3 to work with 4.0.0. used: https://github.com/Darkseal/bootstrap4-glyphicons.
-jquery to 3.3.1. 
-bootstrap3-typeahead updated to work with bootstrap 4
css:
- minor css update to alerts.css, graph.css, prometheus.css. navbar and glyphicon button highlighting color inheritance. 
note: commits show up as nov 5th, as we didn't do commit signoff messages in initial fork. 

cc @juliusv

### CODE REVIEW

**Comment by simonpasquier** [comment]
Thanks for looking into it! You've got conflicts.

File: web/ui/templates/_base.html [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
-        <title>Prometheus Time Series Collection and Processing Server</title>
-        <link rel="shortcut icon" href="{{ pathPrefix }}/static/img/favicon.ico?v={{ buildVersion }}">
-        <script src="{{ pathPrefix }}/static/vendor/js/jquery-3.3.1.min.js?v={{ buildVersion }}"></script>    
-        <script src="{{ pathPrefix }}/static/vendor/bootstrap-4.0.0/js/bootstrap.min.js?v={{ buildVersion }}"></script>
+  <head>
+    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
+    <meta name="robots" content="noindex,nofollow">
+    <title>{{ pageTitle }}</title>
+    <link rel="shortcut icon" href="{{ pathPrefix }}/static/img/favicon.ico?v={{ buildVersion }}">
+    <script src="{{ pathPrefix }}/static/vendor/js/jquery.min.js?v={{ buildVersion }}"></script>
```
Comment (simonpasquier) [inline web/ui/templates/_base.html]: Hmm, are you sure about this line?

File: web/ui/templates/_base.html [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
+    <script src="{{ pathPrefix }}/static/vendor/bootstrap-3.3.1/js/bootstrap.min.js?v={{ buildVersion }}"></script>
```
Comment (simonpasquier) [inline web/ui/templates/_base.html]: Same remark here.

**Comment by simonpasquier** [comment]
Thanks! I'll have a look next week.

File: web/ui/templates/_base.html (hunk 2 above)
Comment (achiuBAE) [reply web/ui/templates/_base.html]: ill take a look at the build problems next week, apologies. its trying to load some bootstrap.scss and also not loading the correct type-ahead file for some reason.

**Comment by achiuBAE** [comment]
Thanks Simon. Hopefully the build passes now, i'll check in next week :)

**Comment by simonpasquier** [comment]
I tried it quickly and it looks good albeit a few remarks. I've tested on Firefox (63.0.1 on Fedora).

* Somehow you're missing https://github.com/prometheus/prometheus/pull/4764.
* There's not enough margin because the top menu bar and the main header.
* Alert rows have no background color.
* The built-in console templates are broken by this change (check `http://<prometheus_URL>/consoles/prometheus.html` for instance).
* The button color is lighter, I don't have a strong opinion on this though so I'm fine keeping the new color.

image

image

**Comment by mkumatag** [comment]
@simonpasquier @achiuBAE Any deadline for this task, if it gets merged then when can we expect this into a release?

**Comment by simonpasquier** [comment]
@mkumatag Prometheus has adopted a time-based release schedule so there should a new Prometheus release every 6 weeks. You can check the schedule [here](https://github.com/prometheus/prometheus/blob/master/RELEASE.md#release-schedule).

@achiuBAE there are new conflicts.

**Comment by ksherryBAE** [comment]
@simonpasquier All conflicts have been resolved, if there is still an issue with the margins clearing your cache should then ensure you're now viewing the latest css.

**Comment by simonpasquier** [comment]
Thanks a lot for doing this! This looks very good. A few minor comments:

* `web/ui/static/vendor/bootstrap-4.1.3/css/.DS_Store` should be removed.
* I still don't see the UI change introduced by https://github.com/prometheus/prometheus/pull/4764.

We're currently in feature freeze waiting for the final v2.6.0 so in any case this PR will have to wait a bit before being merged.

cc @juliusv for feedback.

**Comment by juliusv** [comment]
Couple of things I noticed when playing with this, before looking at code:

- The control buttons in the Graph view are no longer same height and style as their input elements, and no longer connected to them: 
buttons
- In the Graph view's typeahead autocomplete dropdown, you can no longer select elements with KeyUp and KeyDown.
- On the `/targets` page, there is very little spacing between the "All | Unhealthy" selector at the top and the first job group.

**Comment by jritchieBAE** [comment]
> * In the Graph view's typeahead autocomplete dropdown, you can no longer select elements with KeyUp and KeyDown.

@juliusv Two of the three issues you mentioned have been resolved in our latest build. The issue above is due a change in how dropdowns are handled differently in bootstrap4, and there is not a straightforward fix for it. Would you be prepared to accept the build as it is to allow the update to bootstrap4, and maybe have this raised as a seperate new issue if it is something that wants to be added back in?

**Comment by juliusv** [comment]
@jritchieBAE Thanks for the update and the fixes! Unfortunately the arrow key behavior is really fundamental though and not having it would be very frustrating, so I don't think we can afford it being broken. Are there maybe other dropdown/autocomplete libraries that would work with Bootstrap 4? The current one also has the problem that it gets really slow if you have many metric names, see https://github.com/prometheus/prometheus/issues/3321 and https://github.com/prometheus/prometheus/issues/5145. Maybe that would be an opportunity to improve that as well, although that is of course not required for just the Bootstrap update (it should just at least not get worse).

**Comment by simonpasquier** [comment]
Hopefully the autocomplete slowness should be fixed by https://github.com/prometheus/prometheus/pull/5139 which is in good shape from my POV.

**Comment by juliusv** [comment]
@jritchieBAE Just gave this an actual try and there's still some kinks here and there (spacing, styling), but I guess that stuff can be easily fixed. The missing autosuggest keyboard support is an actual problem of course. I'm currently also rebuilding the query browser page in React (with Bootstrap 4 already, so I can base it on this PR), so maybe I will find a dropdown component that works well. Or let me know if you do. I've tried out a couple already, but some are clearly not up for the task (slow with many items) or need a lot of custom work to behave like the one we had.

**Comment by jritchieBAE** [comment]
@juliusv I managed to narrow the issue down. The behaviour of keyboard arrows in the typeahead was being interfered with by the jquery.hotkeys plugin. Disabling this plugin makes the typeahead work as desired. As far as I can tell there are actually no references in the code to the hotkeys file so no functionality is lost.

**Comment by juliusv** [comment]
@jritchieBAE Great, the keyUp/keyDown works now, but I don't see the currently selected dropdown item being highlighted anymore. This makes it hard to see where in the list you are currently navigating. Do you think this is something that can be added back with this autosuggest picker?

**Comment by jritchieBAE** [comment]
@juliusv I'll look into it, hopefully just a CSS change

**Comment by juliusv** [comment]
@jritchieBAE Great. After that, could you rebase on the latest master and rebuild the assets (there's currently a conflict)? And would you be ok with re-opening this PR against a `bootstrap4` branch instead of `master` so I can do some last visual tweaks on it before it goes out?

**Comment by juliusv** [comment]
@jritchieBAE With your latest changes (when rebuilding assets) I get the correct item highlighting now! There is still one odd behavior: if you select one of the items with the cursor keys and press Return to select it, the input field loses focus. This does not happen in current Prometheus, where you want to keep typing in the expression box after selecting a metric (it keeps the focus in the expression box).

**Comment by jritchieBAE** [comment]
> @jritchieBAE With your latest changes (when rebuilding assets) I get the correct item highlighting now! There is still one odd behavior: if you select one of the items with the cursor keys and press Return to select it, the input field loses focus. This does not happen in current Prometheus, where you want to keep typing in the expression box after selecting a metric (it keeps the focus in the expression box).

@juliusv I have been unable to reproduce that issue - for me the input field retains focus. I have tested on multiple browsers.
Also, if you create an empty branch for me to make the new pull request into, I will do so :+1:

**Comment by juliusv** [comment]
@jritchieBAE Hm weird, the loss of focus only happens the second time I select something from the dropdown. E.g.:

1. Go to empty /graph page.
2. Type `node_cpu`.
3. With cursor keys, select `node_cpu_seconds_total` in the dropdown -> focus remains
4. Erase the expression input.
5. Do the same thing again -> no focus on the input field anymore.

I created a new `bootstrap4` branch from `master` against which you should be able to reopen the PR.

**Comment by jritchieBAE** [comment]
In light of the new pull request #5192 , could this pull request be closed?

**Comment by juliusv** [comment]
Yes, thanks, closing here.

**Pull request closed without merging**